ledger/accountdb.go

```diff
@@ -633,6 +644,406 @@ func accountsAddNormalizedBalance(tx *sql.Tx, proto config.ConsensusParams) erro
+	ba.VoteKeyDilution = ad.VoteKeyDilution
+	ba.AuthAddr = ad.AuthAddr
+	ba.TotalAppSchemaNumUint = ad.TotalAppSchema.NumUint
+	ba.TotalAppSchemaNumByteSlice = ad.TotalAppSchema.NumByteSlice
+	ba.TotalExtraAppPages = ad.TotalExtraAppPages
+}
+
+func (ba *baseAccountData) GetAccountData() basics.AccountData {
+	return basics.AccountData{
+		Status:             ba.Status,
+		MicroAlgos:         ba.MicroAlgos,
+		RewardsBase:        ba.RewardsBase,
+		RewardedMicroAlgos: ba.RewardedMicroAlgos,
+		VoteID:             ba.VoteID,
+		SelectionID:        ba.SelectionID,
+		VoteFirstValid:     ba.VoteFirstValid,
+		VoteLastValid:      ba.VoteLastValid,
+		VoteKeyDilution:    ba.VoteKeyDilution,
+		AuthAddr:           ba.AuthAddr,
+		TotalAppSchema: basics.StateSchema{
+			NumUint:      ba.TotalAppSchemaNumUint,
+			NumByteSlice: ba.TotalAppSchemaNumByteSlice,
+		},
+		TotalExtraAppPages: ba.TotalExtraAppPages,
+	}
+}
+
+type resourceFlags uint8
+
+const (
+	resourceFlagsHolding    = 0 //nolint:deadcode
+	resourceFlagsNotHolding = 1
+	resourceFlagsOwnership  = 2
+)
+
+type resourcesData struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	// asset parameters ( basics.AssetParams )
+	Total         uint64         `codec:"a"`
+	Decimals      uint32         `codec:"b"`
+	DefaultFrozen bool           `codec:"c"`
+	UnitName      string         `codec:"d"`
+	AssetName     string         `codec:"e"`
+	URL           string         `codec:"f"`
+	MetadataHash  [32]byte       `codec:"g"`
+	Manager       basics.Address `codec:"h"`
+	Reserve       basics.Address `codec:"i"`
+	Freeze        basics.Address `codec:"j"`
+	Clawback      basics.Address `codec:"k"`
+
+	// asset holding ( basics.AssetHolding )
+	Amount uint64 `codec:"l"`
+	Frozen bool   `codec:"m"`
+
+	// application local state ( basics.AppLocalState )
+	SchemaNumUint      uint64              `codec:"n"`
+	SchemaNumByteSlice uint64              `codec:"o"`
+	KeyValue           basics.TealKeyValue `codec:"p"`
+
+	// application global params ( basics.AppParams )
+	ApprovalProgram               []byte              `codec:"q,allocbound=config.MaxAvailableAppProgramLen"`
+	ClearStateProgram             []byte              `codec:"r,allocbound=config.MaxAvailableAppProgramLen"`
+	GlobalState                   basics.TealKeyValue `codec:"s"`
+	LocalStateSchemaNumUint       uint64              `codec:"t"`
+	LocalStateSchemaNumByteSlice  uint64              `codec:"u"`
+	GlobalStateSchemaNumUint      uint64              `codec:"v"`
+	GlobalStateSchemaNumByteSlice uint64              `codec:"w"`
+	ExtraProgramPages             uint32              `codec:"x"`
+
+	// ResourceFlags helps to identify which portions of this structure should be used; in particular, it
+	// helps to provide a marker - i.e. whether the account was, for instance, opted-in for the asset compared
+	// to just being the owner of the asset. A comparison against the emptry structure doesn't work here -
```

Comment:
emptry -> empty.

Btw, basically only AppLocalState is a problem when local schema is empty

Reply:
it's also an issue for assets. We can create an empty asset where all the params are at their default value. At that point, we won't be able to tell if this account is the owner or not.